trino/trino.go

```diff
@@ -1642,6 +1730,62 @@ func handleResponseError(status int, respErr ErrTrino) error {
 	}
 }
 
+func (qr *driverRows) startSegmentProcessor() {
+	go func() {
+		defer close(qr.stmt.spoolingRowsChannel)
+		defer close(qr.stmt.spoolingProcesserDone)
+		defer close(qr.stmt.downloadPermitCh)
+		buffer := make(map[int64][]queryData)
+		var nextExpectedOffset int64 = 0
+		maxBacklog := 2 * qr.stmt.spoolingMaxOutOfOrderSegments
+		backlogCount := 0
+
+		for {
+			var downloadPermitCh <-chan struct{}
+			if backlogCount <= maxBacklog {
+				downloadPermitCh = qr.stmt.downloadPermitCh
+			} else {
+				downloadPermitCh = nil
+			}
+
+			select {
+			case segment, ok := <-qr.stmt.decodedSegment:
+				if !ok {
+					return
+				}
+
+				buffer[segment.rowOffset] = segment.queryData
+				backlogCount++
+
+				// Try flushing buffered in-order segments
+				for {
+					data, exists := buffer[nextExpectedOffset]
+					if !exists {
+						break
+					}
+
+					select {
+					case qr.stmt.spoolingRowsChannel <- data:
+					case <-qr.doneCh:
+						return
+					}
+
+					delete(buffer, nextExpectedOffset)
+					nextExpectedOffset += int64(len(data))
```

Comment:
We need a way to detect when this is going to get stuck because of a server error, when it'll send data with a gap. It should be possible to check that all downloaders are blocked, all decoders finished, so no more segments will arrive here, and we still don't have the next expected offset.

We need a test that will simulate this.

Reply:
We need to handle a case where the server returns out-of-order data. If we know we can issue a maximum number of N download permits, it means there can only be N-1 out of order segments. If the last segment received here still doesn't have the expected offset, we should fail the whole query, instead of waiting for it to time out.

Reply:
It's enough to check if the buffer is full. It has a fixed size and should never grow. If this happens, it means the server is returning incorrect data, and we should terminate with an error.